src/libproc_macro/lib.rs

```diff
@@ -710,10 +710,11 @@ impl !Sync for Punct {}
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
 #[stable(feature = "proc_macro_12", since = "1.28.0")]
 pub enum Spacing {
-    /// e.g. `+` is `Alone` in `+ =`, `+ident` or `+()`.
+    /// E.g. `+` is `Alone` in `+ =`, `+ident` or `+()`.
     #[stable(feature = "proc_macro_12", since = "1.28.0")]
     Alone,
-    /// e.g. `+` is `Joint` in `+=` or `+#`.
+    /// E.g. `+` is `Joint` in `+=` or `'#`.
+    /// Additionally, single quote `'` can join with identifiers to form lifetimes `'ident`.
```

Comment:
At first I started reporting `Joint` for any non-whitespace token immediately following a punctuation character (for token streams coming from lexer/libsyntax), but it felt wrong enough to revert the change.
It's quite strange to give proc macros ability to discern between e.g. `+x` and `+ x`, something that the "proper language" strives to never do, and combining anything with e.g. doc comments (`+/// Doc`) is plain insane.
So I made a very targeted tweak instead and now report non-punctuation `Joint` only for `'`s followed by identifiers.

Since jointness is more "revealed" than "reported", I believe we'll be able to backward-compatibly "reveal more" in the future if it becomes necessary.